pkg/provenance/provenance_test.go

```diff
@@ -0,0 +1,521 @@
+
+	return a
+}
+// This method is meant to deep copy the Args datastructure.
+// copier deep copies fields that are simple, but I had
+// to deep copy maps and slices myself.
+func deepCopyArgs(arg Args) Args {
+	nArgs := Args{}
+	err := copier.Copy(&nArgs, &arg)
+	if err != nil {
+		fmt.Printf("Error doing a deep copy of Args: %s", err)
+		return nArgs
+	}
+
+	//deep copying for maps,
+	// tried two popular libraries and they didn't work..
+	newUsers := make([]map[string]string, 0)
+	for _, m := range nArgs.Users {
+		u := make(map[string]string, 0)
+		for k, v := range m {
+			u[k] = v
+		}
+		newUsers = append(newUsers, u)
+
+	}
+	newDb := make([]string, 0)
+	for _, m := range nArgs.Databases {
+		newDb = append(newDb, m)
+	}
+	//The above code successfully made a deep copy of the map
+	nArgs.Users = newUsers
+	nArgs.Databases = newDb
+	return nArgs
+}
+//This method builds some lineage data
+func buildLineage() (ObjectLineage, Args) {
+	objectLineage := make(map[int]Spec)
+	args := initArgs()
+	spec1 := makeSpec(args)
+	objectLineage[spec1.Version] = spec1
+
+	args = deepCopyArgs(args)
+	users2 := addUser(args.Users, "johnson", "fluffy23")
+	args.Users = users2
+	args.Version = 2
+	spec2 := makeSpec(args)
+	objectLineage[spec2.Version] = spec2
+
+	args = deepCopyArgs(args)
+	users3 := addUser(args.Users, "thomas", "bedsheet85")
+	args.Users = users3
+	args.Version = 3
+	spec3 := makeSpec(args)
+	objectLineage[spec3.Version] = spec3
+
+	args = deepCopyArgs(args)
+	db1 := addDatabase(args.Databases, "logging")
+	args.Databases = db1
+	args.Version = 4
+	spec4 := makeSpec(args)
+	objectLineage[spec4.Version] = spec4
+
+	args = deepCopyArgs(args)
+	db2 := addDatabase(args.Databases, "demographics")
+	args.Databases = db2
+	args.Version = 5
+	spec5 := makeSpec(args)
+	objectLineage[spec5.Version] = spec5
+	args = deepCopyArgs(args)
+	return objectLineage, args
+}
+
+// This method is testing that the joint functionality of Bisect works
+// with a complex query involving a complex query of 3 fields
+func TestBisectGeneral(t *testing.T) {
+	objLineage, _ := buildLineage()
+	argMapTest := make(map[string]string, 0)
+	argMapTest["field1"] = "databases"
+	argMapTest["value1"] = "demographics"
+	argMapTest["field2"] = "username"
+	argMapTest["value2"] = "daniel"
+	argMapTest["field3"] = "password"
+	argMapTest["value3"] = "4557832+#^"
+
+	vOutput := objLineage.Bisect(argMapTest)
+	expected := "Version: 5"
+	if vOutput != expected {
+		t.Errorf("Version output for TestBisectGeneral() was incorrect, got: %s, want: %s.\n", vOutput, expected)
+	}
+}
+// This method changes a user's password, and is testing
+// whether the bisect will take a query with the same
+// parent element (user), and process a complicated
+// query involving maps.
+func TestBisectPasswordChanged(t *testing.T) {
+	objLineage, newArgs := buildLineage()
+
+	users := updateUser(newArgs.Users, "daniel", "JEK873BUL!")
+	db := removeDatabase(newArgs.Databases, "logging")
```

Comment:
Why removeDatabase in this test?

We can only keep this test for PasswordChange.